CleverRaven/Cataclysm-DDA
# Feature branch powerbanks

#### Summary
Features "Powerbanks now exist"

-----------------------------------------------
#### Purpose of change

The purpose is to add 2 new powerbank items to the game. Powerbanks are relatively common and so it was weird that we have no way of charging phones or other UPS items out in the field without a military grade UPS. These items have been added with a ratio of 1 powerbank / 3-4 phones in the loot tables. The powerbanks themselves should spawn at a ratio of roughly 3 normal powerbanks / 20 cheap powerbanks>

-----------------------------------------------
#### Describe the solution

##### It adds 2 new items:
- a cheap powerbank which is the aquivalent of a 20000 mAh 
- a normal powerbank which is the aquivalent of a 36000 mAh 

##### how it works:
These powerbanks have a charge like normal batteries as opposed to working with a magazine well that can be filled with batteries, as does the UPS. This means these charges cannot be refilled unless the powerbank is hooked up to a functioning power network or placed in a battery charger.

When activating the battery, the player is given 2 options:
- turn on / off : if turned on, the powerbank will start charging other UPS devices. This is equivalent to hooking up a cable between powerbank and a phone irl. The name is left as is to avoid confusion
- plug in / manage cables: similar to other powered tools and is used to charge the powerbank if no battery charger is available.

##### spawns:
I added powerbanks to most locations where a smartphone could be found with a ratio of roughly 1 to 3. This includes locations as well as zombie drops.

##### technicalities:
the weight, volume and max length are based on the 2 powerbanks I own.
The charge capacity is based on a quick guide I found wherein was stated that 20000 mAh powerbanks could charge a phone 4 times, while 36000 mAh powerbanks could charge a phone 6 times. As such, the charges are 4 * 120 and 6 * 120 with 120 being the charge of the smartphone in game.

#### note:
as of yet, the powerbank will not light up yellow in the inventory when turned on. This is an issue with the IS_UPS tag that is need to charge other UPS devices and is also present in the current UPS item. 3/4 of the time spent on this was figuring this out and to fix the Issue I'd have to do a proper deep dive into how the IS_UPS tag affects the active property.

#### Describe alternatives you've considered



-----------------------------------------------

#### Testing

It should not cause any regression since this is a pure addition of new items. No existing code was touched and in the tests I performed I did not run into any issues.

Testing was done as follows:
1. start a new world with a new survivor
2. use debug to spawn in the cheap and normal powerbanks
![afbeelding](https://github.com/CleverRaven/Cataclysm-DDA/assets/145038845/7ddbd5da-219d-43c1-a43b-0d0435397d8b)

3.  Hook them up to an ad-hoc power network to see if they charge 
![afbeelding](https://github.com/CleverRaven/Cataclysm-DDA/assets/145038845/22d901d1-985b-47e0-a5c3-bf5bfc6fee4a)

4.Put them in a battery box to see if they charge 
![afbeelding](https://github.com/CleverRaven/Cataclysm-DDA/assets/145038845/dc19b458-e22b-4417-8f8f-859ac7a7c8f5)

![afbeelding](https://github.com/CleverRaven/Cataclysm-DDA/assets/145038845/1f766b8d-7f39-4a2f-8db3-ae7c6a593dfa)


5.debug in a phone and reduce charge to 0
![afbeelding](https://github.com/CleverRaven/Cataclysm-DDA/assets/145038845/596988c0-517c-444c-9ac6-995dea0487f1)

6. try turning on both powerbanks one by one to see if they charge the phone
![afbeelding](https://github.com/CleverRaven/Cataclysm-DDA/assets/145038845/b2cc30f2-854b-4f69-9f78-700eee0e1048)

![afbeelding](https://github.com/CleverRaven/Cataclysm-DDA/assets/145038845/1beed491-884f-4118-911d-f6e1dec325ae)

7. Now that item functionality was successfully tested, I teleported to the nearest city and debug killed a lot of zombies. The cheap powerbank was found a couple corpses. Both powerbanks are included in a distribution group and only this group is ever used to add powerbanks to loot tables. So I considered this test a success.
